cloudinit/distros/networking.py

```diff
@@ -102,10 +106,69 @@ def is_vlan(self, devname: DeviceName) -> bool:
     def master_is_bridge_or_bond(self, devname: DeviceName) -> bool:
         return net.master_is_bridge_or_bond(devname)
 
+    @abc.abstractmethod
+    def settle(self, *, exists=None) -> None:
```

Comment:
I don't have all the context regarding the creation of that abstraction, but my concern about it is that since this is an abstract method, we cannot ensure that the classes that implement it will follow the `exists` logic detailed in the description here.

Maybe we can enforce that by turning `settle` into this concrete implementation:

```python
def settle(self, *, exists=None) -> None
    if exists is not None:
        exists = self.find_device_path(exists)
    self.settle_device(exists)
```

Then we would need to make both `find_device_path` and `settle_device` abstract. However, I am basing that only on what I saw on the `LinuxNetworking` implementation, so I don't know if that is a pattern we really want to enforce.

Reply:
So in the Linux case, we're calling `udevadm settle` via `util.udevadm_settle`: https://github.com/canonical/cloud-init/blob/2b727914e8cbee6810b1bb9a1cfdb90ad521ceb6/cloudinit/util.py#L2628-L2639

As you can see, `exists` is essentially an optimisation: it allows us to return sooner than we might otherwise, but it doesn't materially change the action being performed. I've updated the docstring to try and reflect this more accurately. Is that a reasonable resolution?

(As an aside: `BSDNetworking` implements this as a `pass` because there is no equivalent on BSD, so this is really Linux-specific behaviour we're talking about: this implementation will only vary if we have a distro that doesn't use udev, which I infer we don't[0] because this code is run unconditionally today.)

[0] Currently; I wonder if Alpine Linux uses udev.

Reply:
@OddBloke Yes, agreed

Reply:
alpine uses [`eudev`](https://pkgs.alpinelinux.org/contents?file=&path=&name=eudev&branch=edge&repo=main&arch=x86) which is `udev` compatible